design/phase-level-progress-tracking.md

```diff
@@ -0,0 +1,191 @@
+
+```
+status:
+    actions:
+    - artifacts:
+        mysqlCloudDump:
+          keyValue:
+            s3path: '{{ .Phases.dumpToObjectStore.Output.s3path }}'
+      blueprint: mysql-blueprint
+      name: backup
+      object:
+        kind: statefulset
+        name: mysql
+        namespace: mysql
+      phases:
+      - name: updatePermissions
+        state: completed
+        progress:
+          progressPercent: "100.00"
+          lastTransitionTime:  "2023-02-20T12:448:55Z"
+      - name: dumpToObjectStore
+        state: running
+        progress:
+          progressPercent: "30.00"
+          lastTransitionTime:  "2023-02-20T12:49:55Z"
+          sizeUploaded: "50000"
+          estimatedTimeSeconds: "120"
+          estimatedUploadSize: "100000000"
+      - name: cleanup
+        state: pending
+    error:
+      message: ""
+    progress:
+      lastTransitionTime: "2023-02-20T12:49:55Z"
+      percentCompleted: "43.33"
+      runningPhase: dumpToObjectStore
+    state: running
+```
+
+### Kanister functions progress tracking
+
+On high level we can divide the Kanister functions into two groups
+
+### Datamover Kanister functions
+
+Kanister data mover functions (like BackupData, CopyVolumeData, RestoreData, etc) use Kopia to snapshot filesystem and move data to/from objects.
+
+While snapshotting, kopia provides the info about progress on the terminal
+
+```
+$ kopia --log-level=error --config-file=/tmp/kopia-repository.config --log-dir=/tmp/kopia-log snapshot --progress-update-interval=5s /data
+ / 1 hashing, 118 hashed (546.9 MB), 0 cached (0 B), uploaded 0 B, estimated 4.5 GB (12.1%) 35s left
+ * 0 hashing, 136 hashed (1.4 GB), 0 cached (0 B), uploaded 244 B, estimated 4.5 GB (30.1%) 30s left
+```
+
+The command output can be parsed to get the progress metadata.
+
+### Non-datamover Kanister functions
+
+Non-datamover Kanister functions like `KubeTask`, `KubeExec`, `KubeOps`, `ScaleWorkload`, etc allow users to perform operations like executing scripts on a Pod or managing K8s resources. The duration it takes to execute these functions depends on different factors like the type of operations, function arguments, and types of commands listed in BP in the case of KubeExec or KubeTask functions. We can roughly divide these function execution into 3 steps.
+
+- Prerequisites - which include steps to perform before running actual operations like setting up env, preparing Pod specs, creating the K8s resources and waiting for them to be ready.
+- Execution - This is the step where the function performs operations.
+- Cleanup - Operations like deleting pods or cleaning up resources
+
+Since there is no standard way to check the progress of these functions, we can divide the progress equally into 3 giving each step equal weightage i.e 33.33%. E.g once the prerequisite step is completed, the progress can be set to 33.33%. And 66.66% once the specified operations are completed.
```

Comment:
```suggestion
Since there is no standard way to check the progress of these functions, we can
divide the progress equally into 3 giving each step equal weightage, i.e.,
33.33%. For example, once the prerequisite step is completed, the progress can
be set to 33.33%, and 66.66% once the specified operations are completed.
```